synapse/storage/data_stores/main/events_worker.py

```diff
@@ -1358,6 +1358,67 @@ def get_next_event_to_expire_txn(txn):
             desc="get_next_event_to_expire", func=get_next_event_to_expire_txn
         )
 
+    @cached(tree=True)
+    async def get_unread_message_count_for_user(
+        self, room_id: str, user_id: str,
+    ):
+        last_read_event_id = await self.get_last_receipt_event_id_for_user(
+            user_id=user_id, room_id=room_id, receipt_type="m.read",
+        )
```

Comment:
you can specify that whenever the cache on `get_last_receipt_event_id_for_user` is invalidated, you want the cache on `get_unread_message_count_for_user` to be invalidated too:

```suggestion
    @cached(tree=True, cache_context=True)
    async def get_unread_message_count_for_user(
        self, room_id: str, user_id: str, cache_context: _CacheContext,
    ):
        last_read_event_id = await self.get_last_receipt_event_id_for_user(
            user_id=user_id, room_id=room_id, receipt_type="m.read",
            on_invalidate=cache_context.invalidate,
        )
```

I think that means that you can remove some of the explicit cache invalidation requests, and might solve your cache invalidation problem?

Reply:
Oh I didn't know that! That might solve it yes, thanks!